File: js/src/simple/concurrent/concurrent.ts
/* eslint-disable no-console */
export function sleep(ms: number) {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, ms);
  });
}

function routine(isCancelled: () => boolean, job: number) {
  return new Promise<void>(async (resolve) => {
    if (isCancelled()) {
      console.log('cancelled: ', job);
      return resolve();
    }
    console.log('work: ', job);
    await sleep(1000);
    console.log('done: ', job);
    return resolve();
  });
}

function processor(
  isCancelled: () => boolean,
  cancel: () => void,
  jobs: number[],
) {
  console.time('processor');
  let promises: Promise<void>[] = [];

  jobs.forEach((job, i) => {
    if (i === 10) cancel();
    // what happens with await below? :)
    const promise = routine(isCancelled, job);
    promises.push(promise);
  });

  return Promise.all(promises).then(() => console.timeEnd('processor'));
}

export async function main() {
  const jobs = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17];
  let _cancelled = false;
  const cancel = () => {
    _cancelled = true;
  };
  const getCancel = () => _cancelled;
  await processor(getCancel, cancel, jobs);
}

main();

File: go/simple/concurrent/concurrent.go
package concurrent

import (
	"context"
	"fmt"
	"sync"
	"time"
)

const CONCURENCY_LIMIT = 10

var wg sync.WaitGroup

// bchannel is a bounded channel allowing up to 10 concurrent operations.
var bchannel = make(chan int, CONCURENCY_LIMIT)

func routine(ctx context.Context, job interface{}) {

	select {
	case <-ctx.Done():
		fmt.Printf("%#v: cancelled\n", job)
	default:
		fmt.Printf("%#v: work\n", job)
		// do work
		time.Sleep(time.Duration(1 * time.Second))
		fmt.Printf("%#v: done\n", job)
	}

	<-bchannel // unblock loop
	wg.Done()  // unblock main
}

func Processor(ctx context.Context, cancel context.CancelFunc, jobs []int) {
	t1 := time.Now()
	for i, job := range jobs {
		bchannel <- 1 // blocks loop if len(bchannel) == CONCURENCY_LIMIT
		wg.Add(1)     // blocks main
		if i == 10 {
			cancel() // last 6 jobs wont run
		}
		go routine(ctx, job)
	}

	wg.Wait()
	elapsed := time.Since(t1).String()
	fmt.Printf("processed %d jobs in %s\n", len(jobs), elapsed)
}
